### PR TITLE
Fix some tensorflow warnings in notebooks

### PR DESCRIPTION
**Pull request by kieranricardo**
This PR fixes some of the warnings in the CI tested demo notebooks

**Tensorflow warnings fixed:**

- ensemble-link-prediction-example has a warning due to changing of names of metrics in tensorflow.

         `WARNING:tensorflow:Early stopping conditioned on metric `val_weighted_acc` which is not 
          available. Available metrics are: loss,acc,val_loss,val_acc`

**Other warnings fixed:**

- sklearn warning about logistic regression in attri2vec-citeseer-node-classification-example.ipynb. The logistic regression model doesn't converge here. This is fixed by increasing the number of iters

       `ConvergenceWarning: lbfgs failed to converge (status=1):`

- pandas warning in cluster-gcn-node-classification.ipynb :

      `SettingWithCopyWarning: A value is trying to be set on a copy of a slice from a DataFrame`

See https://github.com/stellargraph/stellargraph/issues/933 for more info

### CODE REVIEW

**Comment by huonw** [comment]
FWIW, from a developer/review/future perspective, it's much better to be specific about what a patch fixes in the PR description, especially in this case because the comment in #933 discusses more than just what this PR touches, and, the actual changes here are disguised by all the noise from rerunning the notebooks.

The PR description would then go into the squash-and-merge commit message to make it easy to find and useful for someone in future using `git blame`. (For example, in https://github.com/stellargraph/stellargraph/commit/be92b67a42f6a9c5dfcafa4f36015920c1d50ee8 I deleted the auto-generated GitHub message which is noisy/useless, and copied in the description from the PR https://github.com/stellargraph/stellargraph/pull/865 .)

File: demos/ensembles/ensemble-link-prediction-example.ipynb [hunk 1]
```diff
@@ -321,6 +321,7 @@
       "\n",
```
Comment (kjun9) [inline demos/ensembles/ensemble-link-prediction-example.ipynb]: <p>Are we keeping the <code>['...']</code>s  for now?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1008/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='50'/>

File: demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb [hunk 2]
```diff
@@ -104,6 +104,7 @@
       "\n",
```
Comment (kjun9) [inline demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb]: <p>Wow, so much faster!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1008/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (kjun9) [inline demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb]: <p>Comment should be updated to say 20% and 80% if we require 20% to avoid warnings</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1008/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

File: demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb [hunk 3]
```diff
@@ -358,12 +358,12 @@
      "data": {
```
Comment (kjun9) [inline demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb]: <p>Why is the previous cell removed?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1008/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (kieranricardo) [reply demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb]: the previous cell wasn't actually affecting anything downstream and was editing copies of a DataFrame causing the warnings! So I just removed it

File: demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb (hunk 2 above)
Comment (kieranricardo) [reply demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb]: oh I didn't notice; is that from the new stellar class?

File: demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb (hunk 3 above)
Comment (kjun9) [reply demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb]: Hmm I wonder if it's considered bad data science practice to not separate the ground truth labels from the data though?  Is there a way we can still do that without producing warnings?

File: demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb (hunk 2 above)
Comment (kjun9) [reply demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb]: I'm guessing so, each epoch has dropped to 4s from 30s!

File: demos/ensembles/ensemble-link-prediction-example.ipynb (hunk 1 above)
Comment (kieranricardo) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: oh I thought we had fixed the stripping of these warnings?

Yeah keeping them for now because it seems like an issue with tensorflow, e.g. tensorflow's official keras train example has these warnings (https://www.tensorflow.org/guide/keras/train_and_evaluate). I saw someone post a work around of SO but it doesn't worth the effort to implement for all the demos when hopefully this will be fixed in a future release.

File: demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb (hunk 2 above)
Comment (kieranricardo) [reply demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb]: wow, thats amazing! nice work @kjun9 @huonw

Comment (kieranricardo) [reply demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb]: nice catch!

File: demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb (hunk 3 above)
Comment (kieranricardo) [reply demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb]: fair point, it should be pretty easy to to do. i'll give it a go

Comment (kieranricardo) [reply demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb]: done! I added the cell back in but used `train_data = train_data.drop(..., inplace=False, ...)`

File: demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb (hunk 2 above)
Comment (huonw) [reply demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb]: <p>Interesting. I'll investigate to verify, because if it is that much faster, that's a good-news-story we can tell!</p>

File: demos/ensembles/ensemble-link-prediction-example.ipynb (hunk 1 above)
Comment (kjun9) [inline demos/ensembles/ensemble-link-prediction-example.ipynb]: <p>I'm missing a bit of context for why this had <code>val_weighted_acc</code>  in the first place, when it apparently doesn't exist. Possibly worth checking with @PantelisElinas?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1008/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

File: demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb [hunk 4]
```diff
@@ -357,13 +357,13 @@
     {
```
Comment (kjun9) [inline demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb]: <p>Missing a plot here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1008/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='79'/>

Comment (kieranricardo) [reply demos/node-classification/cluster-gcn/cluster-gcn-node-classification.ipynb]: ah.. dunno how that happened but nice catch!

File: demos/ensembles/ensemble-link-prediction-example.ipynb (hunk 1 above)
Comment (kieranricardo) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: yeah I'm not sure but it seems like a typo and I think this warning is new so maybe it was never picked up. Either way I think using the `val_acc` is reasonable here

Comment (kjun9) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: okay sure 👍

Comment (kjun9) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: Hmm I don't see these early stopping parameters in the tensorflow docs for `Model` at all actually 🤔 https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit_generator
or even in the 1.15 docs https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/keras/Model#fit_generator

Does it seem like it's actually doing anything currently...?

The early stopping docs for tensorflow looks very different: https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping

Comment (kieranricardo) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: okay checked this out and just realised that `model` here is an instance of our custom `BaggingEnsemble` not `tf.keras.Model`, and `BaggingEnsemble.fit_generator` supports this argument

Comment (kjun9) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: Oops thanks for figuring that out! 

I do feel like this is potentially a bug though - in an earlier cell it looks like when we compile the model, we use
```python
model.compile(
    optimizer=keras.optimizers.Adam(lr=1e-3),
    loss=keras.losses.binary_crossentropy,
    weighted_metrics=["acc"],
)
```
so `val_weighted_acc` should be available, but it's not? And if we use `val_acc` instead, are we supposed to be passing in `metrics=["acc"]` when we compile the model?

The existing warning message said:
```
WARNING:tensorflow:Early stopping conditioned on metric `val_weighted_acc` which is not available. Available metrics are: loss,acc,val_loss,val_acc
```

so it looks to me like something strange seems to be going on, where the `weighted_metrics` parameter ends up creating a normal metric maybe..? I'm happy for it to be a separate issue though, since the warnings have been removed.

Comment (huonw) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: Why have all the leftover `['...']`s appeared now? Those left-over pieces from the warning removal (#826) weren't there before.

> hopefully this will be fixed in a future release

At the moment, this seems more hopeful than not, as it doesn't seem anyone has filed an issue against tensorflow https://github.com/tensorflow/tensorflow/issues?q=is%3Aissue+is%3Aopen+sample_weight+modes+were+coerced+from . Maybe we can do it; it should be ~easy, since the example in their documentation serves as a reproducer.

Comment (kieranricardo) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: Seems like `tf` is still using weighted accuracy but just isn't adding `"weighted_"` to the string.

Comment (kieranricardo) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: My guess is that these warnings appeared here because this is the first time this notebook has been run with `tensorflow` 2.1.0. 

I'll file an issue with tensorflow and see what they say

File: demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb (hunk 2 above)
Comment (huonw) [reply demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb]: Ah, unfortunately, I cannot reproduce this: I tried running on the latest develop (69fc13bd) and observed epochs taking ~26s, which is the same as what it took on the commit that switched over to the new StellarGraph (2dfff02d701922ed1f2164945dbd75651f9ab6a8), and on the commit before that (172edad0).

Maybe you're running this on a GPU, @kieranricardo?

Comment (huonw) [reply demos/node-classification/attri2vec/attri2vec-citeseer-node-classification-example.ipynb]: > Maybe you're running this on a GPU, @kieranricardo? Or a particularly powerful CPU? (Or some other reason for the difference?)
 
We suspect it may be because he installed `tensorflow-mkl`, and is thus getting the most out of his (Intel) CPU.

File: demos/ensembles/ensemble-link-prediction-example.ipynb (hunk 1 above)
Comment (kieranricardo) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: filed an issue with tensorflow: https://github.com/tensorflow/tensorflow/issues/37500

wow they have a lot of open issues!

Comment (kjun9) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: Yeah looks like it's probably a "feature"
```
>>> import tensorflow as tf
>>> x_in = tf.keras.layers.Input(1)
>>> x_out = tf.keras.layers.Dense(1)(x_in)
>>> model = tf.keras.Model(inputs=x_in, outputs=x_out)
>>> model.compile(
...     optimizer='adam',
...     loss='categorical_crossentropy',
...     weighted_metrics=["acc"],
...     metrics=["acc"]
... )
>>> print(model.metrics_names)
['loss', 'acc', 'acc_1']
```

Comment (kieranricardo) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: @huonw @kjun9 I got a response. Turns out the warning has already been fixed and the `tf-nightly` version doesn't have the warning

Comment (huonw) [reply demos/ensembles/ensemble-link-prediction-example.ipynb]: Great. I wonder if that fix is included in the 2.2.0 release that seems like it'll happen soon (https://github.com/tensorflow/tensorflow/releases/tag/v2.2.0-rc0).
